Taiko

smoogipoo opened the pull request:
This is nearly feature complete and should be feature complete within 16 hours of this PR. PRing for feedback.

Notes:
* osu! mode will not work due to a change in PlayField (removal of ScaleContainer).
* TaikoComboCouner was updated, but is and will remain unused following a discussion in #osu-client.

smoogipoo committed:
Move osu! mode HP display to OsuScoreOverlay.

smoogipoo committed:
Initial work on playfield/hitobject layout.

smoogipoo committed:
Very primitive hit target.

smoogipoo committed:
Update framework.

smoogipoo committed:
Add combo counter and hitcircles to testcase.

smoogipoo committed:
Update framework and resources.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into taiko

Conflicts:
	osu.Game.Modes.Osu/UI/OsuScoreOverlay.cs
	osu.Game/Modes/UI/ScoreOverlay.cs

smoogipoo committed:
Fix post-merge errors.

smoogipoo committed:
Rewrite Taiko hitobjects, add finishers, add basic scoring.

smoogipoo committed:
More work towards a playable game mode.

* Rename classes to Drawable*.
* Add support for parsing Taiko beatmaps.
* Add support for converting osu beatmaps.
* Remove previous (old) classes.

smoogipoo committed:
Fix compile errors + make hitcircles always display for now.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into taiko

smoogipoo committed:
Update framework.

smoogipoo committed:
Add a bit more glow.

smoogipoo committed:
Implement DrumRoll design + adjust triangles pieces.

smoogipoo committed:
Fix incorrect sizing.

smoogipoo committed:
Implement drum roll ticks.

smoogipoo committed:
Minor normal note glow adjustment.

smoogipoo committed:
Add a new test case for hitobjects.

smoogipoo committed:
Remove test case + some code (need to rethink this).

smoogipoo committed:
Implement normal note + finisher judgements.

smoogipoo committed:
Implement hit judgements for everything except spinners. WHERE ARE MY SPINNERS FLYTE?!?!

smoogipoo committed:
Implement some very basic note movement.

smoogipoo committed:
Better drum roll tick positioning.

smoogipoo committed:
Implement scroll speed.

smoogipoo committed:
Add very basic flashy and explody pieces.

smoogipoo committed:
Add basic hit animations to drum roll ticks.

smoogipoo committed:
Center-align hitobjects.

smoogipoo committed:
Initial work towards a more workable playfield.

smoogipoo committed:
Improve scrolling.

smoogipoo committed:
Half implement spinners.

smoogipoo committed:
Fix setting incorrect judgement.

smoogipoo committed:
Fix finisher judgements + input.

smoogipoo committed:
Lower resolution of taiko sprites.

smoogipoo committed:
Re-implement playfield.

smoogipoo committed:
Don't handle repeat keys.

smoogipoo committed:
Split out playfield from visual test.

smoogipoo committed:
Fix incorrect drum roll size.

smoogipoo committed:
Make drum rolls stop at hit target and scale down.

smoogipoo committed:
Adjust hit animations of hitcircles.

smoogipoo committed:
Fix drum roll ticks.

smoogipoo committed:
Revert back to drum rolls moving past the hit target + fix an anchor/origin offset.

smoogipoo committed:
Implement and finish spinners (now called bashes).

smoogipoo committed:
Few minor fixed + derive Playfield.

smoogipoo committed:
Get things sort of working in player.

smoogipoo committed:
Rewrite everything to do with hit effects, plus add kiai time.

smoogipoo committed:
Fix a few animations + add auto hitting.

smoogipoo committed:
Add judgement text.

smoogipoo committed:
Fix spinner inner ring.

smoogipoo committed:
Add basic combo counter.

smoogipoo committed:
Remove custom combocounter, fix drum rolls not containing enough ticks, fix hitobject speed + length.

smoogipoo committed:
Add a few more Beatmap helpers. fix BeatLengthAt not returning velocity-adjusted length, fix Slider.

smoogipoo committed:
Rewrite taiko HitObjects + adjust playfield to be 128-relative.

smoogipoo committed:
Port conversion from sliders to hitcircles in some cases.

smoogipoo committed:
Initialize HitRenderer beatmap in constructor for further extending HitRenderer + disallowing changing the objects.

smoogipoo committed:
Add initial basis for barlines.

smoogipooo commented:
Plz note some things are messy, I haven't had a chance to look through the PR myself yet.

smoogipoo committed:
Bar lines are in.

smoogipoo committed:
Implement hit windows + taiko score processor.

smoogipoo committed:
HP implementation + fixes.

smoogipoo committed:
Cleanup.

Tom94 commented on osu.Game.Modes.Osu/Objects/Slider.cs:
```diff
@@ -53,14 +53,10 @@ public override void SetDefaultsFromBeatmap(Beatmap beatmap)
-            Velocity = baseVelocity / timingPoint.BeatLength;
-            TickDistance = baseVelocity / baseDifficulty.SliderTickRate;
+            Velocity = baseVelocity / beatmap.BeatLengthAt(StartTime);
+            TickDistance = baseVelocity / baseDifficulty.SliderTickRate / beatmap.BPMMultiplierAt(StartTime);
```
Don't do double divides. Order of execution is confusing in such cases. Put parentheses or transform it into `bla / (bla2 * bla3)`

Tom94 commented on osu.Game.Modes.Taiko/Objects/BarLine.cs:
```diff
@@ -0,0 +1,26 @@
+        public void SetDefaultsFromBeatmap(Beatmap beatmap)
+        {
+            // Don't ask... Old osu! had a random multiplier here, that we now have to multiply everywhere
+            float fudgeFactor = 1.4f;
```
Can we do this as part of hitobject / beatmap conversion for the mode, only for legacy beatmaps?

smoogipooo replied:
Nope. This is not limited to convert beatmaps.

smoogipooo replied:
With that being said I have been talking to ppy and we decided that hitobjects should never be passed the beatmap as they are in here. So this method will probably be gone sometime really soon.

Tom94 replied:
I'm aware it's not limited to convert beatmaps. However, I thought hit objects would be passed through Converter regardlessly. Standard osu! hit objects are.

smoogipooo replied:
Good point, though I'll leave it here for now until I/we remove the SetDefaultsFromBeatmap method altogether.

Tom94 commented on osu.Game.Modes.Taiko/Objects/Bash.cs:
```diff
@@ -0,0 +1,28 @@
+            // Todo: Diff range
+            float spinnerRotationRatio = 5;
+
+            RequiredHits = (int)Math.Max(1, Length / 1000f * spinnerRotationRatio * 1.65f);
```
Is 1.65 another random factor? If so, maybe put it into a variable or add a comment.

Tom94 commented on osu.Game.Modes.Taiko/Objects/TaikoHitObject.cs:
```diff
@@ -0,0 +1,54 @@
+            base.SetDefaultsFromBeatmap(beatmap);
+
+            // Don't ask... Old osu! had a random multiplier here, that we now have to multiply everywhere
+            float fudgeFactor = 1.4f;
```
You can make all these fudge factors const. Might also want to share them across multiple files. Seems weird to locally declare them in multiple places.

Tom94 commented on osu.Game.Modes.Taiko/Objects/Drawables/Pieces/RingPiece.cs:
```diff
@@ -0,0 +1,111 @@
+using System.Threading.Tasks;
+
+namespace osu.Game.Modes.Taiko.Objects.Drawables.Pieces
+{
```
For all the "piece" classes you define, it'd be good if you added xml doc explaining what piece exactly they are. Can potentially also do that at the Drawable that contains all the pieces where its children are constructed.

Tom94 commented on osu.Game.Modes.Taiko/UI/TaikoComboCounter.cs:
```diff
@@ -1,63 +1,142 @@
 {
     /// <summary>
-    /// Allows tint and scaling animations. Used in osu!taiko.
+    /// Uses the 'x' symbol and has a pop-out effect while rolling over. Used in osu! standard.
```
"Used in osu! standard"? Why is it named TaikoComboCounter then?

smoogipooo replied:
This... Should be removed lol, the derived ScoreOverlays in general...

Tom94 commented on osu.Game.Modes.Taiko/UI/TaikoHitRenderer.cs:
```diff
@@ -3,17 +3,140 @@
+        {
+            if (h is HitCircle)
+            {
+                switch (h.Type)
```
h.Type should capture all of these. Avoid `is` whereever possible since casting (or checking for validity of a cast) is way more expensive than virtual method calls. In this case it's also less elegant, because we already have the virtual `h.Type`.

Tom94 commented on osu.Game.Modes.Taiko/UI/TaikoPlayfield.cs:
```diff
@@ -8,32 +8,217 @@
+                Position = new Vector2(left_area_size, 0),
+                Size = new Vector2(1f - left_area_size, PLAYFIELD_HEIGHT),
+
+                BorderColour = new Color4(17, 17, 17, 255),
```
Most of the inline colours should probably be in `OsuColour`.

Tom94 commented on osu.Game/Beatmaps/Beatmap.cs:
```diff
@@ -26,11 +27,26 @@ public double BPMAt(double time)
+            return overridePoint?.VelocityAdjustment ?? 1;
+        }
+
+        public double BeatLengthAt(double time, bool addVelocity = true)
```
`multiplyVelocity` if anything. That said, I dislike this method. I'd prefer if you completely left the velocity multiplication out, then added a second helper method which just returns the result of this method and multiplies it by the velocity adjustment.

smoogipooo replied:
This is likely to be the default use case, with multilyVelocity = true. I would go the other way around, or perhaps make it local to the use cases where it's == false. Fair point about the naming though.

Tom94 replied:
I don't see how what kind of parameter should be the default is affecting my proposal. Could you elaborate?

smoogipooo replied:
My point was more leaning towards the second suggestion - that the parameter should be removed altogether and velocity multiplied implicitly. But yes you're right the first suggestion does not affect your proposal.

Tom94 commented on osu.Game/Beatmaps/Beatmap.cs:
```diff
@@ -50,12 +66,66 @@ public ControlPoint TimingPointAt(double time, out ControlPoint overridePoint)
+        /// <param name="hardRockFactor">Factor by which HardRock increases difficulty.</param>
+        /// <param name="mods">Mods to be applied.</param>
+        /// <returns>Modified difficulty value.</returns>
+        public double ApplyModsToDifficulty(double difficulty, double hardRockFactor, Mods mods)
```
This should be static.

Tom94 commented on osu.Game/Beatmaps/Beatmap.cs:
```diff
@@ -50,12 +66,66 @@ public ControlPoint TimingPointAt(double time, out ControlPoint overridePoint)
+        /// <param name="mid">Midpoint of the resulting range which will be achieved by a difficulty value of 5.</param>
+        /// <param name="max">Maximum of the resulting range which will be achieved by a difficulty value of 10.</param>
+        /// <returns>Value to which the difficulty value maps in the specified range.</returns>
+        public double MapDifficultyRange(double difficulty, double min, double mid, double max, Mods mods)
```
So should this be.

Tom94 commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -117,8 +117,12 @@ private void load(AudioManager audio, BeatmapDatabase beatmaps, OsuGameBase game
             ruleset = Ruleset.GetRuleset(usablePlayMode);
 
+            scoreProcessor = ruleset.CreateScoreProcessor(beatmap.HitObjects.Count);
+            scoreProcessor.Initialize(beatmap);
```
Why is initialization and resetting not done inside the score processor constructor? In most cases `Initialize` methods are baaaaad, and an initial `Reset` should be automatic. In general it's a good principle to always make objects be in a valid and usable state after construction.

Tom94 commented:
General note: You forgot license headers in all new files you created. Also, gotta add more xml-doc to public methods of your classes (ideally also protected). We should at some point _require_ all PR'd code to be fully documented. It makes a huge difference for understanding older code that is not fresh in memory.

smoogipoo committed:
Use parentheses instead of double dvision.

smoogipoo committed:
Make slider fudge factor const.

smoogipoo committed:
Remove derived combo counters (all designs will be unified going forward).

smoogipoo committed:
Use enum members for hit object type rather than `is`.

Although I do think this is an uglier method because it has flags, I do understand the reasoning.

smoogipoo committed:
Split out BeatLengthAt into multiple methods.

smoogipoo committed:
Make a few methods static.

smoogipoo committed:
Tidy up ScoreProcessor, make resetting to a stale state implicit.

smoogipoo committed:
Add license headers.

smoogipoo committed:
s/OmitBarLine/OmitFirstBarLine.

smoogipoo committed:
Make barlines fade out after they reach the hit target, before they move under the key area.

smoogipoo committed:
Fix tiiiiiiny inconsistency vs osu-stable resulting in one extra slider tick on some sliders.

smoogipoo committed:
Make non-major barlines slightly more opaque.

smoogipoo committed:
Fix HP and score processor calcs.

smoogipoo committed:
Massive refactoring.

smoogipoo committed:
Offset judgement text by hitobject position.

smoogipoo committed:
Drum roll ticks should not display judgement texts.

smoogipoo committed:
Remove auto.

smoogipoo committed:
More OsuColours.

smoogipoo committed:
Update resources.

smoogipoo committed:
Add gitignore + tracked files.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into taiko

Conflicts:
	osu-framework
	osu-resources
	osu.Desktop.VisualTests/Tests/TestCaseGamefield.cs
	osu.Desktop.VisualTests/Tests/TestCasePlayer.cs
	osu.Game.Modes.Catch/CatchRuleset.cs
	osu.Game.Modes.Catch/UI/CatchHitRenderer.cs
	osu.Game.Modes.Mania/ManiaRuleset.cs
	osu.Game.Modes.Mania/UI/ManiaComboCounter.cs
	osu.Game.Modes.Mania/UI/ManiaHitRenderer.cs
	osu.Game.Modes.Osu/OsuRuleset.cs
	osu.Game.Modes.Osu/OsuScoreProcessor.cs
	osu.Game.Modes.Taiko/Objects/Drawable/DrawableTaikoHit.cs
	osu.Game.Modes.Taiko/Objects/TaikoConverter.cs
	osu.Game.Modes.Taiko/TaikoRuleset.cs
	osu.Game.Modes.Taiko/UI/TaikoComboCounter.cs
	osu.Game.Modes.Taiko/UI/TaikoHitRenderer.cs
	osu.Game.Modes.Taiko/osu.Game.Modes.Taiko.csproj
	osu.Game/Beatmaps/Beatmap.cs
	osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs
	osu.Game/Beatmaps/Timing/ControlPoint.cs
	osu.Game/Modes/Objects/HitObject.cs
	osu.Game/Modes/Ruleset.cs
	osu.Game/Modes/ScoreProcesssor.cs
	osu.Game/Modes/UI/HitRenderer.cs
	osu.Game/Modes/UI/Playfield.cs
	osu.Game/Overlays/Toolbar/Toolbar.cs
	osu.Game/Screens/Play/Player.cs

smoogipoo committed:
Merge remote-tracking branch 'peppy/replay' into taiko

smoogipoo committed:
Adjust colours and add backing glow to drum rolls.

smoogipoo committed:
More adjustments for flyte.

smoogipoo committed:
Add autoplay generation.

smoogipoo committed:
Update resources.

smoogipoo committed:
Restyling.

smoogipoo committed:
Attempt to fix test case.

smoogipoo committed:
Fix TestCaseGamefield.

smoogipooo commented:
This PR is now ready for serious feedback.

smoogipoo committed:
Add a cleaner judgement text animation.

smoogipoo committed:
Animation adjustments to hit target flash.

smoogipoo committed:
Adjust explosion ring animation.

smoogipoo committed:
Adjust hit animations.

smoogipoo committed:
Add missing miss animations.

smoogipoo committed:
Merge remote-tracking branch 'peppy/replay' into taiko

Conflicts:
	osu.Game.Modes.Catch/CatchRuleset.cs
	osu.Game.Modes.Mania/ManiaRuleset.cs
	osu.Game.Modes.Osu/OsuRuleset.cs
	osu.Game.Modes.Taiko/TaikoRuleset.cs
	osu.Game/Modes/UI/HitRenderer.cs

smoogipoo committed:
A bit of refactoring.

smoogipoo committed:
Make taiko autoplay work.

smoogipoo committed:
Adjust input drum animations.

smoogipoo committed:
Minor visual adjustment to hit flash.

smoogipoo committed:
Make replay loading in taiko sort of work (missing some pre-empt offset I think?).

smoogipoo committed:
Make replay loading in taiko sort of work (missing some pre-empt offset I think?).

peppy commented:
- TaikoHitObjects test case is broken
- crashes on completing a map (`TriggerOnJudgement` pushing a results screen twice)

are drum sounds coming in a future PR? how complete is this supposed to be and how much do you want me to point out?

ie. should this be reviewed as a "working to some degree" or "ready for tournament use"?

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/DrawableTaikoHitObject.cs:
```diff
@@ -0,0 +1,68 @@
+
+            // Very naive, but should be enough, given that notes scroll more than 50% of the stage
+            // before start time, so they should scroll off the screen before start time + preempt
+            TaikoHitObject tho = HitObject as TaikoHitObject;
```
unnecessary and non-explicit casts all over this file (don't use `as` unless null checking afterwards)

smoogipooo replied:
Oops, relic of an old past..

smoogipoo committed:
Remove explicit casts.

peppy commented on osu.Game.Modes.Taiko/Objects/DrumRoll.cs:
```diff
@@ -0,0 +1,113 @@
+        {
+            base.SetDefaultsFromBeatmap(beatmap);
+
+            Length *= SLIDER_FUDGE_FACTOR;
```
this call should be non-destructive. this is removing information which cannot be recovered.

smoogipooo replied:
I don't think this is a bad thing. Perhaps the placement of it is not the best, but going forward we should not need to factor in this fudge.

peppy replied:
why is it not done in the beatmap parsing though? if this method is called twice (quite feasible) it will double the fudge factor.

peppy commented on osu.Game.Modes.Taiko/Objects/DrumRoll.cs:
```diff
@@ -0,0 +1,113 @@
+            RequiredGreatHits = (int)(TotalTicks * Math.Min(0.30, 0.10 + (0.20 / 6) * beatmap.BeatmapInfo.BaseDifficulty.OverallDifficulty));
+        }
+
+        public IEnumerable<DrumRollTick> Ticks
```
probably make a method called `CreateTicks()`. very unclear this is instantiating new objects otherwise.

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -0,0 +1,196 @@
+
+namespace osu.Game.Modes.Taiko.Objects.Drawables
+{
+    public class DrawableHitCircleDon : DrawableHitCircle
```
move to own classes, rename to not be don/katsu

smoogipooo replied:
Alright I'll start being pedantic about this stuff again. I was told it would be okay, if you remember that conversation.

peppy replied:
it's okay if:

- it's a nested class
or
- the main class is at the top of the file (right now it's not, leading to confusion
- the main class isn't a large class. in this case it is and would make it hard to find the other two.

peppy replied:
also i wouldn't consider putting multiple classes in one file when they are some of the most important classes in the project

smoogipooo replied:
So what about LegacyReplay? DrawableOsuHitObject? Mod? Heck even Ruleset?
Or do those fall under the "will be fixed soon(tm)" category even though they're already in master?

Is this enforceable? Because maybe it should be, I do agree on enforcing one class/enum/struct per file in absolutely all cases (unless nested).

peppy replied:
most of those should be fixed, yes. possibly with the exception of LegacyReplay as one class could be put inside the other ehre.

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -0,0 +1,196 @@
+
+            double hitOffset = Math.Abs(Judgement.TimeOffset);
+
+            // Must be within great range to be hittable/missable
```
unnecessary (and also incorrect) comment

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -0,0 +1,196 @@
+        protected override void Update()
+        {
+            // If the HitCircle was hit, make it stop moving
+            if (State != ArmedState.Hit)
```
This is dangerous, as it blocks *everything* in the underlying Update. Consider making a property `ShouldMove` or just moving this conditional to the base class. Also I'd move it out to an explicit updateMovement() method.

peppy committed:
Keep x axis and add y axis movement on successful hit.

Also adjusts fade length and removed colour change on miss.

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/DrawableBarLine.cs:
```diff
@@ -0,0 +1,108 @@
+                        Origin = Anchor.TopCentre,
+                        Position = new Vector2(0, -10),
+                        EdgeSmoothness = new Vector2(1),
+                        // Scaling height by 0.866 results in equiangular triangles (== 60° and equal side length)
```
make an EquilateralTriangle class as this is now used in *three* places.

smoogipoo committed:
Structural changes.

smoogipoo committed:
Remove incorrect comment.

smoogipoo committed:
Change location where fudge factor is added to drum roll lengths.

smoogipoo committed:
Fix taiko failing pushing multiple screens.

smoogipoo committed:
A few more comments + refactorings.

smoogipoo committed:
createTicks().

smoogipoo committed:
Add general EquilateralTriangle.

smoogipoo committed:
Fix TestCaseTaikoHitObjects.

smoogipoo committed:
Add mode switching capability to TestCasePlayer.

smoogipoo committed:
Fix future notes being at a lower depth.

smoogipoo committed:
Add fudge factor too non-convert maps.

peppy committed:
Merge remote-tracking branch 'upstream/master' into taiko

peppy committed:
Fix usings.

smoogipoo committed:
Properly abstract fudge factor into TaikoConverter.

smoogipoo committed:
Merge branch 'taiko' of https://github.com/smoogipooo/osu into taiko

peppy committed:
Remove redundant prefixes and sort usings.

peppy committed:
Merge remote-tracking branch 'smoogipooo/taiko' into taiko

peppy committed:
Fix more usings ?__?.

peppy committed:
Add note about function in wrong place.

peppy committed:
Add explicit access prefixes, rename some variables for clarity.

peppy committed:
Avoid possible nullref exception.

peppy committed:
Various fixes.

peppy commented:
- HitObjectConverter.Convert is being run too often.
- private void loadBarLines() shouldn't exist in TaikoHitRenderer (should be managed by the playfield).
- ConvertBarLines() shouldn't be in the converter.
- Shouldn't be using Type or enums all over the place.

peppy commented on osu.Game.Modes.Taiko/Objects/Drawables/Pieces/Circle/CirclePiece.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <summary>
+        /// Whether the HitObject is in kiai time.
+        /// </summary>
+        public bool Kiai;
```
should probably be a property which updates the underlying state

peppy commented:
TaikoConverter -> TaikoHitObjectConverter once we're done.

smoogipoo committed:
Merge commit 'a1aa4ed' into taiko

Conflicts:
	osu.Game.Modes.Taiko/Objects/BarLine.cs
	osu.Game.Modes.Taiko/Objects/Drawables/BarLines/DrawableMajorBarLine.cs

peppy committed:
Make new class TimingInfo to hold all timing related info.

peppy committed:
Refactor the scale factor to be a bit more elegant.

This isn't a permanent solution, but should isolate it to the point where further modifications are menial.

peppy committed:
Merge remote-tracking branch 'smoogipooo/taiko' into taiko

peppy committed:
Fix SMOOGIPOO's usings.

peppy committed:
Improve visual apearance of bar lines.

peppy committed:
Make taiko playfield better.

peppy committed:
Make TimingInfo never null.

peppy committed:
Make taiko hits appear above the left stage when hit well enough.

peppy committed:
Tidy up some unneeded statements and add shadows.

peppy committed:
Use padding instead of margin.

smoogipoo committed:
Drum roll + velocity fixes.

smoogipoo committed:
Ticks should not be altered by velocity.

peppy committed:
Merge branch 'master' into taiko

# Conflicts:
#	osu.Desktop.VisualTests/Tests/TestCasePlayer.cs
#	osu.Game.Modes.Mania/UI/ManiaComboCounter.cs
#	osu.Game/Modes/LegacyReplay.cs
#	osu.Game/Overlays/Toolbar/Toolbar.cs

smoogipoo committed:
Make drum roll tick scoring replicate osu-stable.

peppy committed:
Merge remote-tracking branch 'smoogipooo/taiko' into taiko

peppy committed:
Fix remaining warnings and refactor some.

peppy committed:
Add missed files.

peppy committed:
Further minor clean-ups.
